### PR TITLE
[#10959] Paginate frontend call to webapi for large course

### LINKED ISSUE #10959: Enable large course response download
**Description of feature/enhancement**

Currently, the result download of large courses faces a bottleneck at either the DB access with GAE max timeout 30s or the overall API call with GAE max timeout 60s. Furthermore, too much data being processed also would result in backend OOM. These render instructor unable to download course result whose construction exceeds either of the aforementioned bottlenecks. We should find a scalable and cost effective way to enhance this feature such that it would not have a technical upper bound for the size of the feedback session.

**Justification**

Instructors in large courses with many questions set in one session
Admins who might need to help extract the data for this purpose

**Proposed solution**

We paginate the client request into small chunk with a maximum cap that is guaranteed to not hit the timeout. This comes with an extra cost and hence might need to carefully measure the maximum cap and fit as much data as possible into each paginated call.

### PR DESCRIPTION
**Pull request by moziliar**
Closes #10959

**Outline of Solution**

Break down result download to per question level. Might need to experiment with the threshold to make granular download for smaller courses (OOM issue?)

From the current experiment, since the responses per question fetched are in different servlet, the memory pressure is minimal, possibly due to that GC can safely reclaim the servlet memory. This is better for even smaller courses that might result in memory spike for download (observed for a decently sized course where the memory shot up to close to the maximum)

## Edit

Able to download course far larger than the previous threshold (roughly 600 students, 5 man team and 25 questions) in a synchronous manner, which incur long waiting time that is acceptable.

image

Total memory pressure throughout the download is expected to remain low as compared to the previous implementation where all questions are fetched and bundled together before returning. GC doesn't seem to be working after the servlets return, whose possible cause could be due to memory madvise memory kipnap.

image

Additionally, it is reasonable to believe that the main memory consumption comes from the DB read object. As such, optimizing memory consumption might be hard. Within 12 calls (from the profiler) to fetch 1 question from a 700 students 5 man team course, the heap allocated for `getSessionResultsForUser` reaches 65MB, which averages to about 5MB for a 400KB output. This is reasonable at the moment

image

---

Added a modal that informs the user the progress of the download. This is done in a modal form as the page is effectively frozen while downloading. If the user exits the modal by either closing it or clicking "Abort" button, the download will stop and no csv file will be generated.

## Update

New download modal

108107680-86128f00-70ca-11eb-8f96-ffba0a988bb8

### CODE REVIEW

**Comment by teikjun** [comment]
I didn't notice this earlier, but the "Download Results" button can be accessed from the "Home" and "Sessions" tab via the session table as well. When accessed this way, the download is not processed in batches, so it may TLE like before. I'm still trying to resolve this, I'll merge it together with the frontend changes later. 

image

At the moment, `instructor-session-base-page` doesn't load any question-related information. We'll need to load all the `feedbackQuestionId` (at least), in order to break down the download to question-level. I believe there will be a slight overhead for page loading speed, but I couldn't think of any better solution 😅 

On the UI side, the progress bar is positioned on top of the modal, so it displays correctly on both pages.

**Comment by moziliar** [comment]
@teikjun Sorry that this PR did not tackle that issue... 
I believe solution 1 is more probable. So the rationale for this is that the workflow either consists of
* instructor click download result from session base -> FE fetches all questionIds -> FE downloads with paginated calls
* instructor click view result -> FE fetches all questionIds (and the rest...) -> instructor click download results -> FE downloads with paginated calls
Comparing the above, the former seems more probable as a smoother workflow without additional overhead (or less) in latency. I would encourage that.

**Review by hcwong — APPROVED** [review]
👍 LGTM. The features worked for me - tested with 300 users. Flame graphs and waterfall diagrams were also very helpful. 

However, I did notice that at when there are many students, and we try to load the responses to a question in teammates, it takes quite a while. Perhaps this can also be fixed in another PR.
laggy question fetching

**Comment by moziliar** [comment]
@hcwong Thank you for your review. I figured that improving the performance might incur operational cost. We could by all means parallelise the workflow (pagination again with `forkJoin` or backend forkJoin), but would increase memory consumption and thus instance count. Not sure if the cost is justified.

@damithc prof, any idea if we should give a better UX for this workflow at a cost?

**Comment by damithc** [comment]
> @damithc prof, any idea if we should give a better UX for this workflow at a cost?

Provided the load operation will not time out, perhaps simply show a message below the spinner cautioning the user that this particular question will take a long time to load?

**Comment by moziliar** [comment]
@damithc looks like a good first issue idea. Will just create a ticket? Low priority given no complaints made from the instructors yet.

**Comment by rrtheonlyone** [comment]
Yes that is a good spot @hcwong - let's perhaps create an issue for this and have a separate discussion on this there @moziliar. This PR is more about improving the downloads results feature.

File: src/web/services/simple-modal.service.ts
Comment (rrtheonlyone) [inline src/web/services/simple-modal.service.ts:70]: Maybe rephrase this to `openLoadingModal`? Since this is a generic service that can be reused for other functions.

**Review by rrtheonlyone — CHANGES_REQUESTED** [review]
Awesome work 💯 - I think this PR does a pretty good job of fixing the issues faced in the past when downloading session results for large courses. 

The change is pretty small since we are just leveraging on existing functionality. Good to see some stress testing done above. 

Regarding the user experience:
- We should abort the progress bar automatically upon failure. Right now the progress bar just freezes and we see a toast. 
- There is an empty file committed in the PR (please remove that)
- When we click somewhere on the screen right now, the progress bar disappears but the download still continues to take place - we should not close progress bar modal if user clicks outside.

**Comment by teikjun** [comment]
Thanks for the reviews! The requested changes have been made in https://github.com/moziliar/teammates/pull/23, and will be merged into this main PR soon. I have two concerns:

#### 1. Clicking outside the screen:
The progress bar isn't supposed to disappear when user clicks outside the screen - there is a transparent overlay (behind the loading modal and in front of the content) that is intended to prevent that. I'm not able to reproduce the bug so far on Chrome/Firefox. Can I ask what browser are you using? I'll try to investigate this bug further.

#### 2. Empty files:
I noticed that there are multiple empty `.scss` files in the codebase. Should I open another issue/PR to remove the rest of them?

**Comment by rrtheonlyone** [comment]
I am using chrome version `88.0.4324.192 (Official Build) (x86_64)` on macOS. I just tried and it seems I can click outside and close the modal:

captured_TM

Regarding empty files, small issue - I think can leave it then.

**Comment by teikjun** [comment]
Thanks @rrtheonlyone for the information! The bug has been resolved in moziliar#23. 
We'll review the changes internally and then request a final review from you.

**Review by rrtheonlyone — APPROVED** [review]
LGTM 💯 

UI looks good now! Just one minor thing.. @teikjun could you also abort the download when the 'X' button is pressed in the modal? Right now even after you press the 'X' - the download continues :P Once that is fixed, we can merge in.

**Comment by teikjun** [comment]
The requested changes have been made in https://github.com/moziliar/teammates/pull/24. Good to go after getting approval from @moziliar 👍

**Comment by rrtheonlyone** [comment]
Looks good :)

**Pull request merged**
